### ISSUE TITLE
Spike outputs are downcast from float64 to float32 in multiple snnTorch neuron modules

### ISSUE THREAD

**Issue by GraceDyer** [jeshraghian/snntorch#422]
## Summary

I encountered a reproducible issue where several snnTorch neuron modules produce spike outputs in `float32` even when the inputs (and workflow) are in `float64`.

This appears to break dtype preservation assumptions in high-precision workflows.

## Minimal reproduction

```python
import torch
import snntorch as snn

x = torch.randn(2, 4, dtype=torch.float64)
lif = snn.Leaky(beta=0.9).double()

spk, mem = lif(x)

print(x.dtype)
print(spk.dtype)
```

## Expected behavior

If the module is run with `float64` inputs and converted to `float64`, the output spike tensor should preserve `float64`, or the limitation should be explicitly documented.

## Actual behavior

The input is `torch.float64`, but the spike output is `torch.float32`.

In my reproduction/validation runs, this behavior was stable across repeated reruns and affected multiple modules, including:
- `Leaky`
- `Synaptic`
- `Alpha`
- `Lapicque`
- `RLeaky`
- `RSynaptic`
- `LeakyParallel`
- `SLSTM`

## Environment

- `torch`: `2.8.0`
- `snntorch`: `0.9.4`
- `python`: `3.10`

## Additional notes

This matters for high-precision or numerically sensitive workflows, where users may reasonably expect dtype consistency after converting the model and inputs to `float64`.

If spike outputs are intentionally quantized/downcast internally, an explicit documentation note would be very helpful.

**Comment by tritsystem** [comment]
Thanks for snnTorch! Reporting a reproducible dtype bug with a root cause and a verified fix.

### Describe the bug
When a neuron is driven with `float64` (double-precision) input, the membrane
state correctly stays `float64`, but the **spike output is silently `float32`**.
This breaks double-precision pipelines and quietly mixes dtypes downstream.

### To reproduce
```python
import torch, snntorch as snn

lif = snn.Leaky(beta=0.9)
spk, mem = lif(torch.randn(4, dtype=torch.float64),
               torch.zeros(4, dtype=torch.float64))
print(spk.dtype, mem.dtype)
```

**Expected:** `torch.float64 torch.float64`
**Actual:** `torch.float32 torch.float64`

Also reproduced with `Synaptic` and `Lapicque`.

### Root cause
Every surrogate's `forward` in `snntorch/surrogate.py` hardcodes float32:
```python
out = (input_ > 0).float()   # .float() is always float32, regardless of input dtype
```
The same pattern is in `_surrogate_bypass` in `snntorch/_neurons/neurons.py`.

### Suggested fix
```python
out = (input_ > 0).to(input_.dtype)   # preserve the input dtype
```
Applied to every spike-generating `(input_ > 0).float()`. I verified this fixes
the downcast with **no regression** — float32 behaviour is unchanged and
gradients still flow. Happy to open a PR (bundled with #423 and an `LSO()` fix).

### Environment
- snnTorch **1.0.0** (also reproduced on 0.9.4)
- PyTorch 2.11.0 · Python 3.12

**Comment by ixfd64** [comment]
Should be fixed in #432. Let us know if there are still issues.
